proposals/0000-existentials.rst

```diff
@@ -224,40 +243,54 @@ Proposed Change Specification
 
    #. When checking an expression ``e`` against a type ``C /\ ty``, continue by checking ``e`` against ``ty``,
       emitting ``C`` as a wanted. That is, we must be able to satisfy the constraint ``C`` in order to accept
-      ``e`` as type ``C /\ ty``. (This is reflected in rule ``GenQualified`` in Fig. 11 of the paper_.)
+      ``e`` as type ``C /\ ty``. (This is reflected in rule ``GenQualified`` in Fig. 11 of the paper_.) ::
+
+        Γ ⊩ C ~> dict
+        Γ ⊢ e <= ty ~> e'
+        ----------------------------------------
+        Γ ⊢ e <= C /\ ty ~> Mk/\ @C @ty dict e'
 
    #. When inferring the type of an expression ``e``, if that type is ``exists (a :: k). ty``, then rewrite
       that type to become ``ty[Witness @k (exists (a :: k). ty) e/a]``. (This is rule ``IExist`` in Fig. 5
-      of the paper_.)
+      of the paper_.) ::
+
+        Γ ⊢ e => exists (a :: k). ty ~> e'
+        -------------------------------------------
+        Γ ⊢ e => ty[Witness @k ty e / a] ~> Open e'
 
    #. When inferring the type of an expression ``e``, if that type is ``C /\ ty``, then rewrite that
       type to become ``ty`` and assume ``C`` as a given, available for use anywhere in the innermost
       case-match or lambda, including invisible lambdas as introduced in the argument to a function
       with a higher-rank type. (This is rule ``IGiven`` in Fig. 11 of the paper_.) Use of such a given
       may have a surprising influence on runtime behavior, see Ambiguity_, below.
 
-   #. When inferring the type of a lambda-expression ``\ pats -> e``, we must ensure that no variable bound
-      in ``pats`` leaks into the type of the expression; this is like a skolem-escape.
+        Γ ⊢ e => C /\ ty ~> e'   -- adding C to the environment not modeled here
+        ----------------------
+        Γ ⊢ e => ty ~> sndC e'
 
-      (More technically) When inferring the type of a lambda-expression ``\ pats -> e``: We must bump the
-      type-checker level before checking ``e``. This is because variables bound in ``pats`` might be mentioned
-      in ``Witness``\ es in the type of ``e``. Then, when forming the type of the whole lambda-expression,
-      we promote the inferred type of ``e``; if any variables bound in ``pats`` are encountered, issue an
-      error.
+      See also the ``toVec`` example `below <#toVec>`_.
 
-   #. When trying to satisfy a class constraint ``[W] C (exists (a :: k). ty)``, instead solve
-      ``[W] forall (a :: k). C ty``. For multi-parameter type classes, apply this treatment one
-      argument at a time. Note that this treatment applies to ``~``, which is solved like any
-      other type class (recall that lifted ``~`` is distinct from unlifted ``~#``, to which this
-      treatment does not apply).
+   #. When inferring the type of a ``let`` expression, substitute the bound term variables appearing in
+      the type of the expression with their known right-hand sides. (This is rule ``Let`` in Fig. 4 of the
```

Comment:
How does that work with recursive types? I guess one substitutes the whole letrec… feel free to resolve if there is nothing to worry here

Reply:
No, we have to handle that. I've updated this and added a new bullet around Core typing rules.